### ISSUE TITLE
Keep a history of package transactions

### ISSUE THREAD

**Issue by terrycloth** [hughsie/PackageKit#215]
On Fedora, when I install or remove packages using `dnf`, it keeps a history of these transactions which I can view using the `dnf history` subcommand. I can then undo a specific package installation or removal, or even "rewind" to a previous state (undoing all transactions that happened after the given point in the history).

I think it would be useful if PackageKit were able to do this too. I tend to use GNOME Software only to browse, because it lacks any history tracking.

**Comment by ximion** [comment]
PackageKit does create a history, so this looks like more of a GNOME Software feature request.
Rolling back package upgrades is dangerous, and at least on Debian discouraged, so putting that into PackageKit is something I would not be very happy about.

An update log in GNOME Software could make sense though, but that's for Richard and the GNOME designers to decide.

**Comment by terrycloth** [comment]
> [...] this looks like more of a GNOME Software feature request

Ah, okay, thanks.

> Rolling back package upgrades is dangerous [...]

How so? Fedora doesn't hold onto old versions of packages for long, so depending on how far back you're trying to rewind the system packages, there's the potential that the older versions of the packages you want to revert are no longer available to download --- but that would just mean that the "history undo" command won't be possible. And I imagine Debian is less likely to have that problem.

**Comment by ximion** [comment]
> doesn't seem like the package files or database or anything should end up in an "inconsistent state" or anything

No, but package updates have side effects. For example, and configuration file might be upgraded to a version that only works with a newer version of the package, or directory permissions are changed, system users are added/modified/removed, volatile files are cleaned, caches upgraded etc. Anything that executes maintainer scripts or contains configuration (Debian does a 3-way-merge of those automatically) is generally not safe to roll back to an old version, since we really can't execute these scripts backwards to end up with the previous state (especially not if something gets removed during the update process).
Solutions for this is stateless systems with /etc containing only user overrides, as well as declarative packaging where no scripts are being executed on upgrade/install/removal, and the package manager takes care of that (and also knows how to roll things back from a previous state). The former is being worked on, the latter is a tremendous task, which would generally be nice to have, but at least in Debian isn't being worked on yet - if/when this is addressed, we would still have packages that are not able to be rolled back, but at least the majority of them would be.
tl;dr -> Don't downgrade stuff on Debian systems unless you know what you are doing and what changes the previous update made.

**Comment by kkofler** [comment]
> Fedora doesn't hold onto old versions of packages for long, so depending on how far back you're trying to rewind the system packages, there's the potential that the older versions of the packages you want to revert are no longer available to download.

That's why it is important to keep the cache of downloaded packages by default, so that older versions are always available to downgrade to.
